File: clients/go/page.go
package eoka

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"
)

type Page struct {
	id string
	b  *Browser
}

type MouseButton string

const (
	MouseButtonLeft    MouseButton = "left"
	MouseButtonMiddle  MouseButton = "middle"
	MouseButtonRight   MouseButton = "right"
	MouseButtonBack    MouseButton = "back"
	MouseButtonForward MouseButton = "forward"
)

// CaptchaOptions describes a CAPTCHA challenge to solve in the current browser
// session. The API key and resulting token remain inside the local Eoka server
// process; callers receive only injection metadata.
type CaptchaOptions struct {
	APIKey string `json:"-"`
	// Mode is deliberately required by the server. Explicitly choosing
	// anti_captcha_proxyless prevents a caller from overlooking that the
	// solver runs from a different network than the browser.
	Mode              string          `json:"-"`
	Type              string          `json:"-"`
	WebsiteURL        string          `json:"-"`
	WebsiteKey        string          `json:"-"`
	EnterprisePayload json.RawMessage `json:"-"`
	APIDomain         string          `json:"-"`
	Callback          string          `json:"-"`
}

// CaptchaInjection reports how Eoka applied a solved CAPTCHA token. It never
// contains the token itself.
type CaptchaInjection struct {
	Kind            string   `json:"kind"`
	UpdatedCount    int      `json:"updated_count"`
	Created         []string `json:"created"`
	Callbacks       []string `json:"callbacks"`
	Errors          []string `json:"errors"`
	SolverUserAgent string   `json:"solverUserAgent"`
}

// FetchOptions configures a request executed from the page's browser context.
// Body is sent verbatim; JSON callers should marshal it before calling Fetch.
type FetchOptions struct {
	Method   string
	Headers  map[string]string
	Body     string
	Redirect string
}

// FetchResponse is the result of a browser-context request. Body is returned
// verbatim so callers can decode the API format they requested.
type FetchResponse struct {
	URL     string            `json:"url"`
	Status  int               `json:"status"`
	OK      bool              `json:"ok"`
	Headers map[string]string `json:"headers"`
	Body    string            `json:"body"`
}

func (p *Page) ID() string { return p.id }

func (p *Page) call(ctx context.Context, method string, params map[string]any, result any) error {
	if params == nil {
		params = map[string]any{}
	}
	params["pageId"] = p.id
	return p.b.t.call(ctx, method, params, result)
}

func (p *Page) Goto(ctx context.Context, url string) error {
	return p.call(ctx, "page.goto", map[string]any{"url": url}, nil)
}

func (p *Page) Click(ctx context.Context, selector string) error {
	return p.call(ctx, "page.click", map[string]any{"selector": selector}, nil)
}

// ClickText finds an interactive element by visible text and invokes Eoka's
// semantic element click. This is appropriate for modal continuation controls
// that a site handles through its element listener rather than pointer input.
func (p *Page) ClickText(ctx context.Context, text string) error {
	return p.call(ctx, "page.click_text", map[string]any{"text": text}, nil)
}

func (p *Page) HumanClick(ctx context.Context, selector string) error {
	return p.call(ctx, "page.human_click", map[string]any{"selector": selector}, nil)
}

// HumanClickText finds an interactive element by visible text and clicks it
// with Eoka's human-like pointer input.
func (p *Page) HumanClickText(ctx context.Context, text string) error {
	return p.call(ctx, "page.human_click_text", map[string]any{"text": text}, nil)
}

func (p *Page) Fill(ctx context.Context, selector, value string) error {
	return p.call(ctx, "page.fill", map[string]any{"selector": selector, "value": value}, nil)
}

func (p *Page) HumanFill(ctx context.Context, selector, value string) error {
	return p.call(ctx, "page.human_fill", map[string]any{"selector": selector, "value": value}, nil)
}

func (p *Page) TypeInto(ctx context.Context, selector, text string) error {
	return p.call(ctx, "page.type_into", map[string]any{"selector": selector, "text": text}, nil)
}

func (p *Page) Text(ctx context.Context) (string, error) {
	var res struct {
		Text string `json:"text"`
	}
	err := p.call(ctx, "page.text", nil, &res)
	return res.Text, err
}

func (p *Page) Content(ctx context.Context) (string, error) {
	var res struct {
		HTML string `json:"html"`
	}
	err := p.call(ctx, "page.content", nil, &res)
	return res.HTML, err
}

func (p *Page) Title(ctx context.Context) (string, error) {
	var res struct {
		Title string `json:"title"`
	}
	err := p.call(ctx, "page.title", nil, &res)
	return res.Title, err
}

func (p *Page) URL(ctx context.Context) (string, error) {
	var res struct {
		URL string `json:"url"`
	}
	err := p.call(ctx, "page.url", nil, &res)
	return res.URL, err
}

func (p *Page) GetText(ctx context.Context, selector string) (string, error) {
	var res struct {
		Text string `json:"text"`
	}
	err := p.call(ctx, "page.get_text", map[string]any{"selector": selector}, &res)
	return res.Text, err
}

func (p *Page) GetAttribute(ctx context.Context, selector, name string) (value string, ok bool, err error) {
	var res struct {
		Value *string `json:"value"`
	}
	if err = p.call(ctx, "page.get_attribute", map[string]any{"selector": selector, "name": name}, &res); err != nil {
		return "", false, err
	}
	if res.Value == nil {
		return "", false, nil
	}
	return *res.Value, true, nil
}

func (p *Page) Exists(ctx context.Context, selector string) (bool, error) {
	var res struct {
		Exists bool `json:"exists"`
	}
	err := p.call(ctx, "page.exists", map[string]any{"selector": selector}, &res)
	return res.Exists, err
}

func (p *Page) WaitFor(ctx context.Context, selector string, timeout time.Duration) error {
	return p.call(ctx, "page.wait_for", map[string]any{"selector": selector, "timeoutMs": timeout.Milliseconds()}, nil)
}

func (p *Page) WaitForVisible(ctx context.Context, selector string, timeout time.Duration) error {
	return p.call(ctx, "page.wait_for_visible", map[string]any{"selector": selector, "timeoutMs": timeout.Milliseconds()}, nil)
}

func (p *Page) WaitForText(ctx context.Context, text string, timeout time.Duration) error {
	return p.call(ctx, "page.wait_for_text", map[string]any{"text": text, "timeoutMs": timeout.Milliseconds()}, nil)
}

func (p *Page) Evaluate(ctx context.Context, js string) (json.RawMessage, error) {
	var res struct {
		Result json.RawMessage `json:"result"`
	}
	err := p.call(ctx, "page.evaluate", map[string]any{"js": js}, &res)
	return res.Result, err
}

func EvaluateAs[T any](ctx context.Context, page *Page, js string) (T, error) {
	var out T
	raw, err := page.Evaluate(ctx, js)
	if err != nil {
		return out, err
	}
	if len(raw) == 0 || string(raw) == "null" {
		return out, nil
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return out, fmt.Errorf("eoka: decoding evaluate result: %w", err)
	}
	return out, nil
}

func (p *Page) Execute(ctx context.Context, js string) error {
	return p.call(ctx, "page.execute", map[string]any{"js": js}, nil)
}

// Fetch performs a request in this page's real browser context. It retains the
// page's cookies, browser fingerprint, and same-origin behavior; it is not an
// external HTTP client.
func (p *Page) Fetch(ctx context.Context, url string, options FetchOptions) (FetchResponse, error) {
	params := map[string]any{"url": url}
	if options.Method != "" {
		params["method"] = options.Method
	}
	if len(options.Headers) != 0 {
		params["headers"] = options.Headers
	}
	if options.Body != "" {
		params["body"] = options.Body
	}
	if options.Redirect != "" {
		params["redirect"] = options.Redirect
	}
	var response FetchResponse
	err := p.call(ctx, "page.fetch", params, &response)
	return response, err
}

// SolveCaptcha solves a supported CAPTCHA and injects the result into this
// page. It currently supports recaptcha_v2_enterprise.
func (p *Page) SolveCaptcha(ctx context.Context, options CaptchaOptions) (CaptchaInjection, error) {
	params := map[string]any{
		"apiKey":      options.APIKey,
		"captchaMode": options.Mode,
		"captchaType": options.Type,
		"websiteURL":  options.WebsiteURL,
		"websiteKey":  options.WebsiteKey,
	}
	if len(options.EnterprisePayload) != 0 {
		var payload any
		if err := json.Unmarshal(options.EnterprisePayload, &payload); err != nil {
			return CaptchaInjection{}, fmt.Errorf("eoka: decoding enterprise payload: %w", err)
		}
		params["enterprisePayload"] = payload
	}
	if options.APIDomain != "" {
		params["apiDomain"] = options.APIDomain
	}
	if options.Callback != "" {
		params["callback"] = options.Callback
	}
	var result struct {
		Injection CaptchaInjection `json:"injection"`
	}
	err := p.call(ctx, "page.solve_captcha", params, &result)
	return result.Injection, err
}

func (p *Page) CaptureState(ctx context.Context) (BrowserState, error) {
	var res struct {
		State BrowserState `json:"state"`
	}
	err := p.call(ctx, "page.capture_state", nil, &res)
	return res.State, err
}

func (p *Page) RestoreState(ctx context.Context, state BrowserState) error {
	return p.call(ctx, "page.restore_state", map[string]any{"state": state}, nil)
}

func (p *Page) Screenshot(ctx context.Context) ([]byte, error) {
	var res struct {
		DataBase64 string `json:"dataBase64"`
	}
	if err := p.call(ctx, "page.screenshot", nil, &res); err != nil {
		return nil, err
	}
	data, err := base64.StdEncoding.DecodeString(res.DataBase64)
	if err != nil {
		return nil, fmt.Errorf("eoka: decoding screenshot data: %w", err)
	}
	return data, nil
}

func (p *Page) Select(ctx context.Context, selector, value string) error {
	return p.call(ctx, "page.select", map[string]any{"selector": selector, "value": value}, nil)
}

func (p *Page) Hover(ctx context.Context, selector string) error {
	return p.call(ctx, "page.hover", map[string]any{"selector": selector}, nil)
}

func (p *Page) PressKey(ctx context.Context, key string) error {
	return p.call(ctx, "page.press_key", map[string]any{"key": key}, nil)
}

func (p *Page) MouseDown(ctx context.Context, x, y float64, button MouseButton) error {
	return p.call(ctx, "page.mouse_down", map[string]any{"x": x, "y": y, "button": button}, nil)
}

func (p *Page) MouseMove(ctx context.Context, x, y float64) error {
	return p.call(ctx, "page.mouse_move", map[string]any{"x": x, "y": y}, nil)
}

func (p *Page) MouseUp(ctx context.Context, x, y float64, button MouseButton) error {
	return p.call(ctx, "page.mouse_up", map[string]any{"x": x, "y": y, "button": button}, nil)
}

func (p *Page) KeyDown(ctx context.Context, key string) error {
	return p.call(ctx, "page.key_down", map[string]any{"key": key}, nil)
}

func (p *Page) KeyUp(ctx context.Context, key string) error {
	return p.call(ctx, "page.key_up", map[string]any{"key": key}, nil)
}

func (p *Page) ReleaseAllInputs(ctx context.Context) error {
	return p.call(ctx, "page.release_all_inputs", nil, nil)
}

func (p *Page) Close(ctx context.Context) error {
	return p.call(ctx, "page.close", nil, nil)
}

File: crates/eoka-server/src/methods/page.rs
use base64::Engine;
use captcha::{build_captcha_inject_js, AntiCaptcha, CaptchaInjectionKind};
use serde::Deserialize;
use serde_json::{json, Map, Value};
use std::collections::BTreeMap;

use super::browser::close_tab_impl;
use super::{parse_params, PageIdParams};
use crate::protocol::ServerError;
use crate::state::AppState;

fn core_mouse_button(button: eoka_protocol::MouseButton) -> eoka::MouseButton {
    match button {
        eoka_protocol::MouseButton::Left => eoka::MouseButton::Left,
        eoka_protocol::MouseButton::Middle => eoka::MouseButton::Middle,
        eoka_protocol::MouseButton::Right => eoka::MouseButton::Right,
        eoka_protocol::MouseButton::Back => eoka::MouseButton::Back,
        eoka_protocol::MouseButton::Forward => eoka::MouseButton::Forward,
    }
}

fn single(key: &'static str, value: impl Into<Value>) -> Value {
    let mut map = Map::with_capacity(1);
    map.insert(key.to_string(), value.into());
    Value::Object(map)
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct SelectorParams {
    page_id: String,
    selector: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct TextParams {
    page_id: String,
    text: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct GotoParams {
    page_id: String,
    url: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct FillParams {
    page_id: String,
    selector: String,
    value: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct TypeIntoParams {
    page_id: String,
    selector: String,
    text: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct GetAttributeParams {
    page_id: String,
    selector: String,
    name: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct WaitForParams {
    page_id: String,
    selector: String,
    timeout_ms: u64,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct WaitForTextParams {
    page_id: String,
    text: String,
    timeout_ms: u64,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct JsParams {
    page_id: String,
    js: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct FetchParams {
    page_id: String,
    url: String,
    #[serde(default = "default_fetch_method")]
    method: String,
    #[serde(default)]
    headers: BTreeMap<String, String>,
    #[serde(default)]
    body: Option<String>,
    #[serde(default = "default_fetch_redirect")]
    redirect: String,
}

fn default_fetch_method() -> String {
    "GET".into()
}

fn default_fetch_redirect() -> String {
    "follow".into()
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct PressKeyParams {
    page_id: String,
    key: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct MouseButtonParams {
    page_id: String,
    #[serde(flatten)]
    mouse: eoka_protocol::MouseButtonArgs,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct MouseMoveParams {
    page_id: String,
    #[serde(flatten)]
    mouse: eoka_protocol::MouseMoveArgs,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct HeldKeyParams {
    page_id: String,
    #[serde(flatten)]
    key: eoka_protocol::KeyArgs,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct RestoreStateParams {
    page_id: String,
    state: eoka::BrowserState,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct SolveCaptchaParams {
    page_id: String,
    api_key: String,
    captcha_type: String,
    captcha_mode: String,
    #[serde(rename = "websiteURL")]
    website_url: String,
    website_key: String,
    #[serde(default)]
    enterprise_payload: Option<Value>,
    #[serde(default)]
    api_domain: Option<String>,
    #[serde(default)]
    callback: Option<String>,
}

pub async fn goto(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: GotoParams = parse_params(params)?;
    state.page(&params.page_id)?.goto(&params.url).await?;
    Ok(json!({}))
}

pub async fn click(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: SelectorParams = parse_params(params)?;
    state.page(&params.page_id)?.click(&params.selector).await?;
    Ok(json!({}))
}

pub async fn click_text(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: TextParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .click_by_text(&params.text)
        .await?;
    Ok(json!({}))
}

pub async fn human_click(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: SelectorParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .human_click(&params.selector)
        .await?;
    Ok(json!({}))
}

pub async fn human_click_text(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: TextParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .human_click_by_text(&params.text)
        .await?;
    Ok(json!({}))
}

pub async fn fill(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: FillParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .fill(&params.selector, &params.value)
        .await?;
    Ok(json!({}))
}

pub async fn human_fill(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: FillParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .human_fill(&params.selector, &params.value)
        .await?;
    Ok(json!({}))
}

pub async fn type_into(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: TypeIntoParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .type_into(&params.selector, &params.text)
        .await?;
    Ok(json!({}))
}

pub async fn text(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    let text = state.page(&params.page_id)?.text().await?;
    Ok(single("text", text))
}

pub async fn content(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    let html = state.page(&params.page_id)?.content().await?;
    Ok(single("html", html))
}

pub async fn title(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    let title = state.page(&params.page_id)?.title().await?;
    Ok(single("title", title))
}

pub async fn url(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    let url = state.page(&params.page_id)?.url().await?;
    Ok(single("url", url))
}

pub async fn get_text(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: SelectorParams = parse_params(params)?;
    let page = state.page(&params.page_id)?;
    let element = page.find(&params.selector).await?;
    let text = element.text().await?;
    Ok(single("text", text))
}

pub async fn get_attribute(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: GetAttributeParams = parse_params(params)?;
    let page = state.page(&params.page_id)?;
    let element = page.find(&params.selector).await?;
    let value = element.get_attribute(&params.name).await?;
    Ok(single("value", value))
}

pub async fn exists(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: SelectorParams = parse_params(params)?;
    let exists = state.page(&params.page_id)?.exists(&params.selector).await;
    Ok(json!({ "exists": exists }))
}

pub async fn wait_for(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: WaitForParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .wait_for(&params.selector, params.timeout_ms)
        .await?;
    Ok(json!({}))
}

pub async fn wait_for_visible(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: WaitForParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .wait_for_visible(&params.selector, params.timeout_ms)
        .await?;
    Ok(json!({}))
}

pub async fn wait_for_text(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: WaitForTextParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .wait_for_text(&params.text, params.timeout_ms)
        .await?;
    Ok(json!({}))
}

pub async fn evaluate(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: JsParams = parse_params(params)?;
    let result: Value = state.page(&params.page_id)?.evaluate(&params.js).await?;
    Ok(single("result", result))
}

pub async fn execute(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: JsParams = parse_params(params)?;
    state.page(&params.page_id)?.execute(&params.js).await?;
    Ok(json!({}))
}

/// Performs a request from the current page's browser context.
///
/// Unlike an external HTTP client, this uses the page's live cookies,
/// browser fingerprint, and same-origin policy. The response body is returned
/// verbatim so callers can decode JSON or another API format themselves.
pub async fn fetch(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: FetchParams = parse_params(params)?;
    let script = build_browser_fetch_js(&params)?;
    let body: String = state.page(&params.page_id)?.evaluate(&script).await?;
    serde_json::from_str(&body)
        .map_err(|error| ServerError::internal(format!("invalid browser fetch response: {error}")))
}

fn build_browser_fetch_js(params: &FetchParams) -> Result<String, ServerError> {
    if !matches!(params.redirect.as_str(), "follow" | "error" | "manual") {
        return Err(ServerError::invalid_params(
            "redirect must be follow, error, or manual",
        ));
    }
    let url = serde_json::to_string(&params.url)
        .map_err(|error| ServerError::internal(error.to_string()))?;
    let method = serde_json::to_string(&params.method)
        .map_err(|error| ServerError::internal(error.to_string()))?;
    let headers = serde_json::to_string(&params.headers)
        .map_err(|error| ServerError::internal(error.to_string()))?;
    let body = serde_json::to_string(&params.body)
        .map_err(|error| ServerError::internal(error.to_string()))?;
    let redirect = serde_json::to_string(&params.redirect)
        .map_err(|error| ServerError::internal(error.to_string()))?;

    Ok(format!(
        r#"(async () => {{
            const response = await fetch({url}, {{
                method: {method},
                headers: {headers},
                body: {body},
                credentials: "include",
                redirect: {redirect},
            }});
            const body = await response.text();
            return JSON.stringify({{
                url: response.url,
                status: response.status,
                ok: response.ok,
                headers: Object.fromEntries(response.headers.entries()),
                body,
            }});
        }})()"#
    ))
}

pub async fn capture_state(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    let state = state.page(&params.page_id)?.capture_state().await?;
    let state =
        serde_json::to_value(state).map_err(|error| ServerError::internal(error.to_string()))?;
    Ok(single("state", state))
}

pub async fn restore_state(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: RestoreStateParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .restore_state(&params.state)
        .await?;
    Ok(json!({}))
}

pub async fn screenshot(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    let png = state.page(&params.page_id)?.screenshot().await?;
    let data_base64 = base64::engine::general_purpose::STANDARD.encode(png);
    Ok(single("dataBase64", data_base64))
}

pub async fn select(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: FillParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .select(&params.selector, &params.value)
        .await?;
    Ok(json!({}))
}

pub async fn hover(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: SelectorParams = parse_params(params)?;
    state.page(&params.page_id)?.hover(&params.selector).await?;
    Ok(json!({}))
}

pub async fn press_key(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PressKeyParams = parse_params(params)?;
    state.page(&params.page_id)?.press_key(&params.key).await?;
    Ok(json!({}))
}

pub async fn mouse_down(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: MouseButtonParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .mouse_down(
            params.mouse.x,
            params.mouse.y,
            core_mouse_button(params.mouse.button),
        )
        .await?;
    Ok(json!({}))
}

pub async fn mouse_move(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: MouseMoveParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .mouse_move(params.mouse.x, params.mouse.y)
        .await?;
    Ok(json!({}))
}

pub async fn mouse_up(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: MouseButtonParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .mouse_up(
            params.mouse.x,
            params.mouse.y,
            core_mouse_button(params.mouse.button),
        )
        .await?;
    Ok(json!({}))
}

pub async fn key_down(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: HeldKeyParams = parse_params(params)?;
    state
        .page(&params.page_id)?
        .key_down(&params.key.key)
        .await?;
    Ok(json!({}))
}

pub async fn key_up(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: HeldKeyParams = parse_params(params)?;
    state.page(&params.page_id)?.key_up(&params.key.key).await?;
    Ok(json!({}))
}

pub async fn release_all_inputs(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    state.page(&params.page_id)?.release_all_inputs().await?;
    Ok(json!({}))
}

/// Solve a supported CAPTCHA and apply its token inside the current page.
///
/// The token intentionally never crosses the server protocol boundary. It is
/// both sensitive and only useful in the browser session that requested it.
pub async fn solve_captcha(state: &AppState, params: Value) -> Result<Value, ServerError> {
    let params: SolveCaptchaParams = parse_params(params)?;
    if params.captcha_mode != "anti_captcha_proxyless" {
        return Err(ServerError::invalid_params(
            "captchaMode must explicitly be anti_captcha_proxyless; proxy-backed and manual modes belong to the caller",
        ));
    }
    let solution = match params.captcha_type.as_str() {
        "recaptcha_v2_enterprise" => {
            AntiCaptcha::new(params.api_key)
                .solve_recaptcha_v2_enterprise(
                    &params.website_url,
                    &params.website_key,
                    params.enterprise_payload,
                    params.api_domain.as_deref(),
                )
                .await
        }
        other => {
            return Err(ServerError::invalid_params(format!(
                "unsupported captcha type {other:?}; supported: recaptcha_v2_enterprise"
            )));
        }
    }
    .map_err(|error| ServerError::internal(format!("captcha solve failed: {error}")))?;

    let token = solution
        .token()
        .ok_or_else(|| ServerError::internal("captcha solver returned no token"))?;
    let solver_user_agent = solution.user_agent.clone();
    let script = build_captcha_inject_js(
        token,
        CaptchaInjectionKind::Recaptcha,
        params.callback.as_deref(),
    );
    let page = state.page(&params.page_id)?;
    let mut result = captcha_injection_result(page, &script).await?;
    let callback_count = result
        .get("callbacks")
        .and_then(Value::as_array)
        .map_or(0, Vec::len);
    // reCAPTCHA widgets can attach their callback client just after a token is
    // first delivered. Re-scan once; the injection script de-duplicates
    // already-delivered callbacks by token and path.
    if callback_count <= 1 {
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        let settled = captcha_injection_result(page, &script).await?;
        if settled
            .get("callbacks")
            .and_then(Value::as_array)
            .is_some_and(|callbacks| !callbacks.is_empty())
        {
            result = settled;
        }
    }
    // Preserve non-secret solver metadata for callers that need to diagnose
    // identity consistency. The CAPTCHA token never leaves this process.
    if let Some(user_agent) = solver_user_agent {
        result["solverUserAgent"] = Value::String(user_agent);
    }
    Ok(single("injection", result))
}

async fn captcha_injection_result(page: &eoka::Page, script: &str) -> Result<Value, ServerError> {
    let result: String = page.evaluate(script).await?;
    serde_json::from_str(&result).map_err(|error| {
        ServerError::internal(format!("invalid CAPTCHA injection result: {error}"))
    })
}

pub async fn close(state: &mut AppState, params: Value) -> Result<Value, ServerError> {
    let params: PageIdParams = parse_params(params)?;
    if let Some(error) = close_tab_impl(state, &params.page_id).await? {
        return Err(error.into());
    }
    Ok(json!({}))
}

#[cfg(test)]
mod tests {
    use super::{build_browser_fetch_js, FetchParams};
    use std::collections::BTreeMap;

    #[test]
    fn browser_fetch_uses_page_credentials_and_encodes_values() {
        let script = build_browser_fetch_js(&FetchParams {
            page_id: "page-1".into(),
            url: "https://example.test/api?x=\"quoted\"".into(),
            method: "POST".into(),
            headers: BTreeMap::from([("Content-Type".into(), "application/json".into())]),
            body: Some(r#"{"value":"quoted"}"#.into()),
            redirect: "follow".into(),
        })
        .unwrap();

        assert!(script.contains("credentials: \"include\""));
        assert!(script.contains("\"POST\""));
        assert!(script.contains("\\\"quoted\\\""));
    }

    #[test]
    fn browser_fetch_rejects_unknown_redirect_mode() {
        let error = build_browser_fetch_js(&FetchParams {
            page_id: "page-1".into(),
            url: "https://example.test".into(),
            method: "GET".into(),
            headers: BTreeMap::new(),
            body: None,
            redirect: "somewhere".into(),
        })
        .unwrap_err();

        assert!(error.message.contains("redirect"));
    }
}
